pkg/instancetype/instancetype.go

```diff
@@ -1386,6 +1387,41 @@ func applyInputPreferences(preferenceSpec *instancetypev1beta1.VirtualMachinePre
 	}
 }
 
+func applyPanicDevicePreferences(preferenceSpec *instancetypev1beta1.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) {
+	if len(preferenceSpec.Devices.PreferredPanicDevices) == 0 {
+		return
+	}
+
+	emptyPanicDevice := virtv1.PanicDevice{}
+	foundVMIEmptyPanicDevice := false
+	foundPreferredEmptyPanicDevice := false
+
+	// Only apply any preferred panic device when the same panic device has not been provided by a user already
+	panicDeviceModels := make(map[string]struct{})
+	for _, panicDevice := range vmiSpec.Domain.Devices.PanicDevices {
+		if panicDevice.Model == nil {
+			foundVMIEmptyPanicDevice = true
+			continue
+		}
+		panicDeviceModels[*panicDevice.Model] = struct{}{}
+	}
+
+	for _, preferredPanicDevice := range preferenceSpec.Devices.PreferredPanicDevices {
+		if preferredPanicDevice.Model == nil {
+			foundPreferredEmptyPanicDevice = true
+			continue
+		}
+
+		if _, found := panicDeviceModels[*preferredPanicDevice.Model]; !found {
+			vmiSpec.Domain.Devices.PanicDevices = append(vmiSpec.Domain.Devices.PanicDevices, preferredPanicDevice)
+		}
+	}
+
+	if foundPreferredEmptyPanicDevice && !foundVMIEmptyPanicDevice {
+		vmiSpec.Domain.Devices.PanicDevices = append(vmiSpec.Domain.Devices.PanicDevices, emptyPanicDevice)
+	}
+}
```

Comment:
Would you mind adding some unit tests for this? Appreciate the tests are a mess at the moment but we will be breaking them up during the 1.5 dev cycle finally.

Reply:
I've updated the unit-tests. Also added some api validations to the vmi spec. I'm yet to add validations for the preference spec. I'll do that and update here

Reply:
@lyarwood I've added api validations for the panic device specs.